flink-core/src/main/java/org/apache/flink/api/common/operators/util/SlotSharingGroupUtils.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.common.operators.util;
+
+import org.apache.flink.api.common.operators.ResourceSpec;
+import org.apache.flink.api.common.operators.SlotSharingGroup;
+import org.apache.flink.api.common.resources.ExternalResource;
+import org.apache.flink.util.Preconditions;
+
+import java.util.stream.Collectors;
+
+/** Utils for {@link SlotSharingGroup}. */
+public class SlotSharingGroupUtils {
+    public static ResourceSpec extractResourceSpec(SlotSharingGroup slotSharingGroup) {
+        if (!slotSharingGroup.getCpuCores().isPresent()) {
+            return ResourceSpec.UNKNOWN;
+        }
+
+        Preconditions.checkState(slotSharingGroup.getCpuCores().isPresent());
+        Preconditions.checkState(slotSharingGroup.getTaskHeapMemory().isPresent());
+        Preconditions.checkState(slotSharingGroup.getTaskOffHeapMemory().isPresent());
+        Preconditions.checkState(slotSharingGroup.getManagedMemory().isPresent());
+
+        return ResourceSpec.newBuilder(slotSharingGroup.getCpuCores().get(), 0)
```

Comment:
The unit of this param is `MB`. So, I think we cannot use `slotSharingGroup.getTaskHeapMemory().get()` here.

Reply:
It's looks weird if we do not accept a 0 value taskHeapMemory and make it a required param.
I prefer to add a new factory method `newBuilder()` for `ResourceSpec` and a paramless constructor for the `Builder`, similar to `ResourceProfile`.

Reply:
I would rather add a new factory method `newBuilder(double cpu, MemorySize taskHeapMemory)`. cpu & task-heap memory are always explicitly provided is one of the assumption of `ResourceSpec` as @xintongsong said.

Reply:
Sounds good. But it's better to add a check in the existing and new `newBuilder()` to ensure that `cpu` and `taskHeapMemory` are larger than 0. Otherwise the assumption is possible to be broken by mistake.

Reply:
I do think so. However, as this PR is large enough. I'd like to file a new ticket to address it.

Reply:
Sounds good to me.